benchmarks/tensors/dibaryon/baryon_generator.cpp

```diff
@@ -47,168 +210,329 @@ void generate_function(std::string name)
+      // NOTE
+      // iterators of P = { t, iCprime, iSprime, kCprime, kSprime, x, y },
+      //
+      complex_expr p = p_computation(t, jCprime, jSprime, kCprime, kSprime, x, y);
+
+      complex_expr bdouble_update_def =
+        Bdouble_init(t, n, iCprime, iSprime, jCprime, jSprime, kCprime, kSprime, x, x2) -
+        p * prop(t, 2, iCprime, iSprime, kc, ks, x2, y) * psi;
+      complex_computation bdouble_update(
+          // name
+          str_fmt("bdouble_p_update_%d_%d", kc, ks),
+          // iterator
+          {t, jCprime, jSprime, kCprime, kSprime, x, n, iCprime, iSprime, x2, y},
+          // definition
+          bdouble_update_def);
+      Bdouble_p_updates.push_back(bdouble_update);
+
+      computation *p_real;
+      computation *p_imag;
+      std::tie(p_real, p_imag) = kAndComp.second;
+      P2UserEdge edge {p_real, p_imag, bdouble_update.get_real(), bdouble_update.get_imag()};
+      p2userEdges.push_back(edge);
+    }
 
-    std::pair<expr, expr> prop_2pp(prop_r(t, 2, iCprime, iSprime, lCprime, lSprime, x2, y), prop_i(t, 2, iCprime, iSprime, lCprime, lSprime, x2, y));
-    computation Bdouble_r_update1("Bdouble_r_update1", {t, n, iCprime, iSprime, jCprime, jSprime, kCprime, kSprime, lCprime, lSprime, x, x2, y},
-	    Bdouble_r_init(t, n, iCprime, iSprime, jCprime, jSprime, kCprime, kSprime, x, x2) - mul_r(P_update, prop_2pp));
+    struct O2UserEdge {
+      computation *o_r, *o_i,
+                  *bd_r, *bd_i;
+    };
+    std::vector<O2UserEdge> o2userEdges;
+    for (auto &iAndComp : O) {
+      int ic, is;
+      std::tie(ic, is) = iAndComp.first;
+
+      complex_computation o_computation = iAndComp.second;
+      // 
+      // NOTE
+      // iterators of O { t, iCprime, iSprime, kCprime, kSprime, x, y },
+      //
+      complex_expr o = o_computation(t, jCprime, jSprime, kCprime, kSprime, x, y);
+
+      complex_expr bdouble_update_def =
+        Bdouble_init(t, n, iCprime, iSprime, jCprime, jSprime, kCprime, kSprime, x, x2) +
+        o * prop(t, 0, iCprime, iSprime, ic, is, x2, y) * psi;
+      complex_computation bdouble_update(
+          // name
+          str_fmt("bdouble_o_update_%d_%d", ic, is),
+          // iterator
+          {t, jCprime, jSprime, kCprime, kSprime, x, n, iCprime, iSprime, x2, y},
+          // definition
+          bdouble_update_def);
+
+      Bdouble_o_updates.push_back(bdouble_update);
+
+      computation *o_real;
+      computation *o_imag;
+      std::tie(o_real, o_imag) = iAndComp.second;
+      O2UserEdge edge {o_real, o_imag, bdouble_update.get_real(), bdouble_update.get_imag()};
+      o2userEdges.push_back(edge);
+    }
 
-    computation Bdouble_i_update1("Bdouble_i_update1", {t, n, iCprime, iSprime, jCprime, jSprime, kCprime, kSprime, lCprime, lSprime, x, x2, y},
-	    Bdouble_i_init(t, n, iCprime, iSprime, jCprime, jSprime, kCprime, kSprime, x, x2) - mul_i(P_update, prop_2pp));
+#define Q_BLOCK_SIZE 3
+#define O_BLOCK_SIZE 3
+#define P_BLOCK_SIZE 3
```

Comment:
Can you move all the defines to the beginning of this file? Can you also document what are these used for?